src/ConservationChecker.jl

```diff
@@ -0,0 +1,331 @@
+    end
+
+    # save atmos
+    push!(cc.ρe_tot_atmos, sum(atmos_sim.integrator.u.c.ρe_tot))
+
+    # save land
+    if land_sim !== nothing
+        e_per_area_land = zeros(axes(land_sim.integrator.u.bucket.W))
+        get_land_energy(land_sim, e_per_area_land)
+        push!(cc.ρe_tot_land, sum(e_per_area_land .* surface_masks.land))
+    else
+        push!(cc.ρe_tot_land, FT(0))
+    end
+
+    # save sea ice
+    if ice_sim !== nothing
+        push!(cc.ρe_tot_seaice, sum(get_slab_energy(ice_sim, ice_sim.integrator.u.T_sfc) .* surface_masks.ice))
+    else
+        push!(cc.ρe_tot_seaice, FT(0))
+    end
+
+    # save ocean
+    if ocean_sim !== nothing
+        push!(cc.ρe_tot_ocean, sum(get_slab_energy(ocean_sim, ocean_sim.integrator.u.T_sfc) .* surface_masks.ocean))
+    else
+        push!(cc.ρe_tot_ocean, FT(0))
+    end
+
+    # save surface friction sink
+    include_ke_dissipation = false # NB: currently, kinetic energy dissipation is neglected in total energy tendency
+    if include_ke_dissipation
+        push!(cc.friction_sink, sum((ke_dissipation(atmos_sim)) .* coupler_sim.Δt_cpl)) # ρ d ke_friction / dt 
+    else
+        push!(cc.friction_sink, FT(0))
+    end
+
+    push!(cc.ice_base_source, FT(0))
+
+    tot =
+        @. cc.ρe_tot_atmos + cc.ρe_tot_ocean + cc.ρe_tot_land + cc.ρe_tot_seaice + cc.toa_net_source + cc.friction_sink
+    # @assert abs((tot[end] - tot[1]) / tot[end]) < 1e-4
+end
+
+"""
+    ke_dissipation(sim)
+
+Calculate the kinetic energy diffipation of a simulation. 
+"""
+function ke_dissipation(sim)
+    drag_uv =
+        .-Geometry.UVVector.(
+            Geometry.Covariant12Vector.(
+                sim.integrator.p.dif_flux_uₕ_bc.components.data.:1,
+                sim.integrator.p.dif_flux_uₕ_bc.components.data.:2,
+            )
+        )
+    dot.(Geometry.UVVector.(level(sim.integrator.u.c.uₕ, 1)), drag_uv) .* level(sim.integrator.u.c.ρ, 1)
+end
+
+"""
+        check_conservation(
+        cc::WaterConservationCheck,
+        coupler_sim,
+        )
+
+computes the total water ∫ ρq_tot dV of the various components
+of the coupled simulations, and updates cc with the values.
+
+Note: in the future this should not use `push!`.
+"""
+function check_conservation!(
+    cc::WaterConservationCheck,
+    coupler_sim::CoupledSimulation,
+    get_slab_energy,
+    get_land_energy,
+)
+    @unpack model_sims, surface_masks = coupler_sim
+    @unpack atmos_sim, land_sim, ocean_sim, ice_sim = model_sims
+
+    boundary_space = coupler_sim.boundary_space
+    FT = eltype(coupler_sim.surface_masks.land)
+
+    # save atmos
+    push!(cc.ρq_tot_atmos, sum(atmos_sim.integrator.u.c.ρq_tot))
+
+    # save land
+    if land_sim !== nothing
+        water_content =
+            @. (land_sim.integrator.u.bucket.σS + land_sim.integrator.u.bucket.W + land_sim.integrator.u.bucket.Ws) # m^3 water / land area / layer height
+        parent(water_content) .= parent(water_content .* surface_masks.land) * FT(1000)
```

Comment:
It's a bit hard for me to figure out the unit here (without knowing details about the land model). Maybe add some more comments?

Reply:
Sure! I can comment more lines. Also, instead of the FT(1000), I'm adding  `ClimaLSM.LSMP.ρ_cloud_liq` which clarifies the units further. The final values are in kg, thanks to ClimaCore's `sum` function doing the volume integral for us (rather than a simple sum).